File: pegasusX/apps/factory-portal/lib/manifest-lifecycle.ts
import {
  factoryManifestCompleteKey,
  factoryManifestDispatchKey,
  factoryManifestSealKey,
  factoryManifestStartLoadingKey,
} from '@pegasusx/api-client';

export type ManifestLifecycleAction = {
  path: 'start-loading' | 'seal' | 'dispatch' | 'complete';
  label: string;
};

export function manifestTransitionIdempotencyKey(
  manifestId: string,
  action: ManifestLifecycleAction['path'],
  factoryId = '',
): string {
  switch (action) {
    case 'start-loading':
      return factoryManifestStartLoadingKey(manifestId);
    case 'seal':
      return factoryManifestSealKey(manifestId, factoryId);
    case 'dispatch':
      return factoryManifestDispatchKey(manifestId, factoryId);
    case 'complete':
      return factoryManifestCompleteKey(manifestId, factoryId);
    default:
      return `factory-manifest-transition:${factoryId}:${manifestId}:${action}`;
  }
}

export function nextManifestLifecycleAction(state: string): ManifestLifecycleAction | null {
  switch (state.toUpperCase()) {
    case 'DRAFT':
      return { path: 'start-loading', label: 'Start loading' };
    case 'LOADING':
      return { path: 'seal', label: 'Seal manifest' };
    case 'SEALED':
      return { path: 'dispatch', label: 'Dispatch' };
    case 'DISPATCHED':
      return { path: 'complete', label: 'Complete' };
    default:
      return null;
  }
}

export const MANIFEST_STATE_ORDER = ['DRAFT', 'LOADING', 'SEALED', 'DISPATCHED', 'COMPLETED', 'CANCELLED'] as const;

File: pegasusX/apps/factory-app-android/app/src/main/java/com/pegasusx/factory/ui/screens/manifest/ManifestLifecycle.kt
package com.pegasusx.factory.ui.screens.manifest

import com.pegasusx.factory.data.model.ManifestTransitionRequest
import com.pegasusx.factory.data.remote.FactoryApi
import com.pegasusx.factory.util.FactoryIdempotencyKeys
import retrofit2.Response
import com.pegasusx.factory.data.model.ManifestTransitionResponse

data class ManifestLifecycleStep(
    val path: String,
    val label: String,
)

fun nextManifestLifecycleStep(state: String): ManifestLifecycleStep? = when (state.uppercase()) {
    "DRAFT" -> ManifestLifecycleStep("start-loading", "Start loading")
    "LOADING" -> ManifestLifecycleStep("seal", "Seal manifest")
    "SEALED" -> ManifestLifecycleStep("dispatch", "Dispatch")
    "DISPATCHED" -> ManifestLifecycleStep("complete", "Complete")
    else -> null
}

suspend fun applyManifestLifecycle(
    api: FactoryApi,
    manifestId: String,
    step: ManifestLifecycleStep,
): Response<ManifestTransitionResponse> {
    val body = ManifestTransitionRequest(reason = "factory-app-android")
    val idempotencyKey = FactoryIdempotencyKeys.forLifecyclePath(manifestId, step.path)
    return when (step.path) {
        "start-loading" -> api.startManifestLoading(manifestId, idempotencyKey, body)
        "seal" -> api.sealManifest(manifestId, idempotencyKey, body)
        "dispatch" -> api.dispatchManifest(manifestId, idempotencyKey, body)
        "complete" -> api.completeManifest(manifestId, idempotencyKey, body)
        else -> error("unsupported manifest lifecycle path: ${step.path}")
    }
}
